### ISSUE TITLE
`histplot` isn't showing bar for a bin in certain distributions

### ISSUE THREAD

**Issue by kretes** [mwaskom/seaborn#2748]
For an example like this:
```
import numpy as np
import seaborn as sns

np.random.seed(42)

ZEROS=700
ONES=30000

dummy_data = np.hstack([np.zeros(ZEROS), 1 - (np.random.binomial(1000, 0.1, ONES) / 1000)])

sns.histplot(dummy_data)
```
We get a plot without any visible bar at x==0:
![image](https://user-images.githubusercontent.com/1224887/155480016-5ef7bc1b-18a8-4134-ac44-492062cdc398.png)

Interestingly if I Make the distribution 10 times smaller:
```
ZEROS=70
ONES=3000
```
The bar starts being visible.

Another way to overcome it is to use fill=False:
![image](https://user-images.githubusercontent.com/1224887/155480226-77dbfb72-432f-4699-aa3a-667290b01be5.png)

I have as well noticed that it depends on the figsize, and another way to make the bar visible is to add e.g. `plt.figure(figsize=(9,4))` before:
![image](https://user-images.githubusercontent.com/1224887/155480502-67e1e100-6800-4206-adce-e1bc045ef1c6.png)

See reproduction in collab: https://colab.research.google.com/drive/1x79KLWA6mJrr1-P8zsKNfKCV5EoZ4lhr?usp=sharing

**Comment by mwaskom** [comment]
If you have discrete data and a binwidth <1, some bins will start and end between integer values and not have any observations fall into them.

**Comment by jhncls** [comment]
This is an aliasing effect. Matplotlib filters away narrow bars if they "fall between two pixels". Note that matplotlib treats areas different from lines, which are antialiased (showed dimmed, independent whether they fall between the pixels or not). Matplotlib doesn't antialias areas as that would interfere with alpha compositing of touching areas.

Workarounds:

- explicitly set a linewidth (`sns.histplot(dummy_data, lw=0.2)`)
- explicitly set the number of bins to something smaller (seaborn uses `np.histogram_bin_edges(..., bins='auto')` for the default, which a.o. takes the number of samples into account)
- set a higher dpi when saving (or using a larger figsize)

Related matplotlib issues:

- https://github.com/matplotlib/matplotlib/issues/8808
- https://github.com/matplotlib/matplotlib/issues/13909

**Comment by mwaskom** [comment]
Oh thanks @jhncls i had misunderstood this as being about the gaps in the distribution on the right, but yes aliasing artifact is a good explanation for what’s going on on the left side of the plot.

**Comment by kretes** [comment]
Thanks for looking at this issue. The data is not discrete, but you are right the distribution is quite obscure - but that's what real-world data sometimes is.

What I aim is that the default behavior could be set such that this situation is less likely to influence a user.

You're right that it might be the number of bins that contribute to this - for the dummy data I created it gives `1131`

Maybe a solution would be to use a different/customized method for `bin_edges` - something like set the maximum number of bin edges to the value that will make each bin at least visible, taking into the consideration aliasing that happens afterwards.
I'm sure this is yet another heuristic that won't make seaborn code easier to get, but it might help.

Or change the default of `fill` to `False` - as you can see it works ok with this. If this is a safer solution overall (overcoming matplotlib aliasing effects) - this might be the default?

**Comment by mwaskom** [comment]
> The data is not discrete

Sorry, aren't you generating data from a binomial distribution?

> Maybe a solution would be to use a different/customized method for bin_edges

There's an ongoing [diffusion-of-responsibility](https://github.com/numpy/numpy/issues/15569) between numpy and downstream plotting libraries over what a more robust "auto" histogram would look like. There's no clear defensible solution and changing things in numpy is hard, but I would strongly prefer to defer all computation here to numpy and not have to implement / support / defend an ad hoc approach.

> If this is a safer solution overall (overcoming matplotlib aliasing effects) - this might be the default?

It's extremely difficult to compare layered unfilled histograms or to visually disambiguate between layering and stacking. It's a worse default option if it handles this odd corner-case better. 

FWIW I don't actually get this effect on my system with a default copy-and-paste of your example. It's probably dependent on number of factors. In any case I don't think there is a huge risk of mistaken interpretations since the x axis scale autoscales properly, which should cause one to suspect that something weird is happening. And the seaborn histogram documentation strongly urges the user not to accept the default bin size and to try other options. That applies here too!

**Comment by kretes** [comment]
> > The data is not discrete
> 
> Sorry, aren't you generating data from a binomial distribution?

Right, in the example yes. I was thinking that my original data wasn't discrete and this issue occured.

> 
> > Maybe a solution would be to use a different/customized method for bin_edges
> 
> There's an ongoing [diffusion-of-responsibility](https://github.com/numpy/numpy/issues/15569) between numpy and downstream plotting libraries over what a more robust "auto" histogram would look like. There's no clear defensible solution and changing things in numpy is hard, but I would strongly prefer to defer all computation here to numpy and not have to implement / support / defend an ad hoc approach.

Fair enough, thanks very much for the explanation. It might be, though, that there are different 'auto' strategy applicable for different scenarios and we cannot expect numpy's `auto` be the best for all.

> 
> > If this is a safer solution overall (overcoming matplotlib aliasing effects) - this might be the default?
> 
> It's extremely difficult to compare layered unfilled histograms or to visually disambiguate between layering and stacking. It's a worse default option if it handles this odd corner-case better.
> 
> FWIW I don't actually get this effect on my system with a default copy-and-paste of your example. It's probably dependent on number of factors. In any case I don't think there is a huge risk of mistaken interpretations since the x axis scale autoscales properly, which should cause one to suspect that something weird is happening. And the seaborn histogram documentation strongly urges the user not to accept the default bin size and to try other options. That applies here too!

Right, good point again.

**Comment by mwaskom** [comment]
> we cannot expect numpy's auto be the best for all

Well to be clear *no* reference rule will be "best for all" scenarios — but that discussion is revolving around the idea that "auto" is primarily intended to support plotting usecases.
